### PR TITLE
Add initial documentation for URLPattern.

### PR DESCRIPTION
**Pull request by wanderview**
This adds initial interface documentation for the URLPattern interface being worked on in https://github.com/WICG/urlpattern.

### CODE REVIEW

File: content/en-US/api/URLPattern.md [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+
+// Example interface page at: https://developer.mozilla.org/en-US/docs/Web/API/ContentIndex
+
+## Description
+
+// Complete the sentence below with a basic description of the API.
+// You may add more sentences if needed.
+
+The `URLPattern` interface provides a web platform primitive for matching
+URLs.  The pattern syntax is adopted from the popular [path-to-regexp][]
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md:29]: This is for matching in URLs in service workers, right? Please say that.

File: content/en-US/api/URLPattern.md [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+## Constructor
+
+Creates a new `URLPattern` object.
+
+## Properties
+
+**`URLPattern.protocol`**
+
+Returns the URL protocol pattern set during construction.  This value may
+be differ from the input to the constructor due to normalization.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md:71]: What's normalization? We'll need to cover that somehow, but I need to understand it before I can recommend an appropriate way to discuss it.

File: content/en-US/api/URLPattern.md [hunk 3]
```diff
@@ -0,0 +1,515 @@
...
+// Complete the sentence below with a basic description of the API.
+// You may add more sentences if needed.
+
+The `URLPattern` interface provides a web platform primitive for matching
+URLs.  The pattern syntax is adopted from the popular [path-to-regexp][]
+javascript library and is more ergonomic than using regular expressions.
+
+## Constructor
+
+Creates a new `URLPattern` object.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
Creates a new `URLPattern` object. It takes two arguments.

**`URLPatternInput`**

An object describing the URLs you want to match. Its members can be any of
 `protocol`, `username`, `password`, `hostname`, `port`, `pathname`, `search`,
 or `hash`.

**`baseURL`**

A {{domxref('USVString')}} that ...
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: Constructor will need to contain something like this. I don't imagine what I've suggested is correct because I have questions.

- My suggestion matches the IDL that's currently in Chrome, but it doesn't match the examples in your explainer. Which is correct.
- I'm not able to infer what `baseURL` is or how it interacts with the members of `URLPatternInput`.  I would think that `baseURL` would just be a pattern that you want to match, but your examples resolve it to something literal like `self.origin` or `self.location`.

File: content/en-US/api/URLPattern.md [hunk 4]
```diff
@@ -0,0 +1,515 @@
...
+
+## Properties
+
+**`URLPattern.protocol`**
+
+Returns the URL protocol pattern set during construction.  This value may
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
Returns the URL protocol pattern passed to the constructor.  This value may
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: This is the more common way to say it on MDN and seems to me to be a bit clearer to a web developer since it describes what the developer does, not what the UA does. Please change this throughout.

File: content/en-US/api/URLPattern.md [hunk 5]
```diff
@@ -0,0 +1,515 @@
...
+be differ from the input to the constructor due to normalization.
+
+**`URLPattern.username`**
+
+Returns the URL username pattern set during construction.  This value may
+be differ from the input to the constructor due to normalization.
+
+**`URLPattern.password`**
+
+Returns the URL username pattern set during construction.  This value may
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
Returns the URL password pattern set during construction.  This value may
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: It looks like you forgot to update your cut and paste work. Please fix throughout.

File: content/en-US/api/URLPattern.md [hunk 6]
```diff
@@ -0,0 +1,515 @@
...
+## Events
+
+None
+
+## Methods
+
+**`URLPattern.test()`**
+
+Matches the input against the pattern returning true on success and false
+otherwise.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
Returns a boolean indicating whether the passed string matches the current pattern.
```

File: content/en-US/api/URLPattern.md [hunk 7]
```diff
@@ -0,0 +1,515 @@
...
+
+**`URLPattern.exec()`**
+
+Matches the input against the pattern returning a URLPatternResult object on
+success and null otherwise.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
Returns a `URLPatternResult`  object containing detailed information about the match along with capture groups, or null if there is no match.
```

File: content/en-US/api/URLPattern.md (hunk 6 above)
Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: More consistent with MDN style.

File: content/en-US/api/URLPattern.md [hunk 8]
```diff
@@ -0,0 +1,515 @@
...
+
+## Examples
+
+### Example: Construct a URLPattern to filter on a specific URL component.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Filter on a Specific URL Component
```

File: content/en-US/api/URLPattern.md [hunk 9]
```diff
@@ -0,0 +1,515 @@
...
+
+The following example shows how a URLPattern can be used to filter on a
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
The following example shows how a URLPattern filters a
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: This is an example of a writing best practice to omit needless words: specifically, words that don't contribute meaning to the sentence.

File: content/en-US/api/URLPattern.md [hunk 10]
```diff
@@ -0,0 +1,515 @@
...
+specific URL component.  When the URLPattern constructor is called with a
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
specific URL component.  When the `URLPattern()` constructor is called with a
```

File: content/en-US/api/URLPattern.md [hunk 11]
```diff
@@ -0,0 +1,515 @@
...
+structured object of component patterns any missing components default to
+the `*` wildcard value.
+
+```js
+// Construct a URLPattern that matches a specific domain and its subdomains.
+// All other URL components default to the wildcard `*` pattern.
+const pattern = new URLPattern({
+  hostname: {*.}?example.com
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]: Does the value of `hostname` need to be in quotes? When I run this in a console, the period causes 'SyntaxError: Unexpected token'.

File: content/en-US/api/URLPattern.md [hunk 12]
```diff
@@ -0,0 +1,515 @@
...
+});
+
+// `{*.}?example.com`
+pattern.hostname;
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
// Prints '{*.}?example.com' 
console.log(pattern.hostname);
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: The change to this doesn't need to be as radical as I at first thought. My change is to show you how MDN would handle this. Please change throughout.

File: content/en-US/api/URLPattern.md [hunk 13]
```diff
@@ -0,0 +1,515 @@
...
+pattern.test({ hostname: "cdn.example.com" });
+
+// true
+pattern.test("custom-protocol://example.com/other/path?q=1");
+
+// false because the hostname component does not match
+pattern.test("https://cdn-example.com/foo/bar");
+```
+
+### Example: Construct a URLPattern from a full URL string.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Construct a URLPattern from a Full URL String.
```

File: content/en-US/api/URLPattern.md [hunk 14]
```diff
@@ -0,0 +1,515 @@
...
+
+The following example shows how a URLPattern can be constructed from a full
+URL string with patterns embedded.  If there is no ambiguity between whether
+a character is a URL separate or a pattern character then it just works.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]: What are URL separates and pattern characters?

File: content/en-US/api/URLPattern.md [hunk 15]
```diff
@@ -0,0 +1,515 @@
...
+pattern.hash;
+
+// true
+pattern.test("https://cdn-1234.example.com/product/assets/hero.jpg");
+
+// false because the search component does not match
+pattern.test("https://cdn-1234.example.com/product/assets/hero.jpg?q=1");
+```
+
+### Example: Constructing a URLPattern with an ambiguous URL string.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Constructing a URLPattern with an Ambiguous URL String.
```

File: content/en-US/api/URLPattern.md [hunk 16]
```diff
@@ -0,0 +1,515 @@
...
+
+The following example shows how a URLPattern constructed from an ambiguous
+string will favor treating characters as part of the pattern syntax.  In
+this case the `:` character could be the protocol component suffix or it
+could be the prefix for a pattern named group.  The constructor chooses
+to treat this as part of the pattern and therefore determines this is
+a relative pathname pattern.  Since there is no base URL the relative
+pathname cannot be resolved and it throws.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
pathname cannot be resolved and it throws an error.
```

File: content/en-US/api/URLPattern.md [hunk 17]
```diff
@@ -0,0 +1,515 @@
...
+
+```js
+// Throws because this is interpreted as a single relative pathname pattern
+// with a ":foo" named group and there is no base URL.
+const pattern = new URLPattern("data:foo*");
+```
+
+### Example: Escaping characters to disambiguate URLPattern constructor strings.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Escaping Characters to Disambiguate URLPattern Constructor Strings
```

File: content/en-US/api/URLPattern.md [hunk 18]
```diff
@@ -0,0 +1,515 @@
...
+pattern.hostname;
+pattern.port;
+pattern.search;
+pattern.hash;
+
+// true
+pattern.test("data:foobar");
+```
+
+### Example: Using base URLs for `test()` and `exec()`.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Using Base URLs for test() and exec()
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: Special formatting for code keywords is often avoided in headings. In headings the disambiguation value of such formatting is minuscule and leaving it out makes a page visually more appealing.

File: content/en-US/api/URLPattern.md [hunk 19]
```diff
@@ -0,0 +1,515 @@
...
+
+The following example show how input values to `test()` and `exec()` can use
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
The following example shows how input values to `test()` and `exec()` can use
```

File: content/en-US/api/URLPattern.md [hunk 20]
```diff
@@ -0,0 +1,515 @@
...
+result.pathname.input;
+
+// `bar`
+result.pathname.groups[0];
+
+// `example.com`
+result.hostname.input;
+```
+
+### Example: Using base URLs in the pattern constructor.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Using base URLs in the URLPattern Constructor.
```

File: content/en-US/api/URLPattern.md [hunk 21]
```diff
@@ -0,0 +1,515 @@
...
+
+The follow example shows how base URLs can also be used to construct the
+URLPattern.  Its important to note that the base URL in these cases is
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
URLPattern.  It's important to note that the base URL in these cases is
```

File: content/en-US/api/URLPattern.md [hunk 22]
```diff
@@ -0,0 +1,515 @@
...
+treated strictly as a URL and cannot contain any pattern syntax itself.
+
+Also, since the base URL provides a value for every component the resulting
+URLPattern will also have a value for every component; even if its the
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
URLPattern will also have a value for every component; even if it's the
```

File: content/en-US/api/URLPattern.md [hunk 23]
```diff
@@ -0,0 +1,515 @@
...
+const pattern2 = new URLPattern("/foo/*", "https://example.com" });
+
+// Throws because a relative constructor string must have a base URL to resolve
+// against.
+try {
+  const pattern3 = new URLPattern("/foo/*");
+} catch (e) {}
+```
+
+### Example: Accessing matched group values.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Accessing Matched Group Values
```

File: content/en-US/api/URLPattern.md [hunk 24]
```diff
@@ -0,0 +1,515 @@
...
+result.hostname.groups[0];
+
+// `cdn.example.com`
+result.hostname.input;
+
+// [{ hostname: "cdn.example.com" }]
+result.inputs;
+```
+
+### Example: Accessing matched group values using custom names.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Example: Accessing matched group values using custom names.
```
```suggestion
### Accessing Matched Group Values Using Custom Names
```

File: content/en-US/api/URLPattern.md [hunk 25]
```diff
@@ -0,0 +1,515 @@
...
+result.pathname.groups.action;
+
+// `/store/wanderview/view`
+result.pathname.input;
+
+// [{ pathname: "/store/wanderview/view" }]
+result.inputs;
+```
+
+### Example: Custom regular expression groups.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Custom Regular Expression Groups
```

File: content/en-US/api/URLPattern.md [hunk 26]
```diff
@@ -0,0 +1,515 @@
...
+// false
+pattern.test({ pathname: "/baz" });
+
+const result = pattern.exec({ pathname: "/foo" });
+
+// `foo`
+result.pathname.groups[0];
+```
+
+### Example: Named group with a custom regular expression.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Named Group with a Custom Regular Expression
```

File: content/en-US/api/URLPattern.md [hunk 27]
```diff
@@ -0,0 +1,515 @@
...
+
+```js
+const pattern = new URLPattern({ pathname: "/:type(foo|bar)" });
+const result = pattern.exec({ pathname: "/foo" });
+
+// `foo`
+result.pathname.groups.type;
+```
+
+### Example: Making matching groups optional.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Making Matching Groups Optional
```

File: content/en-US/api/URLPattern.md [hunk 28]
```diff
@@ -0,0 +1,515 @@
...
+pattern3.test({ pathname: "/product/wanderview/view" });
+
+// true
+pattern3.test({ pathname: "/product" });
+
+// true
+pattern3.test({ pathname: "/product/" });
+```
+
+### Example: Making matching groups repeated.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Making Matching Groups Repeated
```

File: content/en-US/api/URLPattern.md [hunk 29]
```diff
@@ -0,0 +1,515 @@
...
+
+The following example shows how a matching group can be made repeated by placing
+a `+` modifier after it.  Again, in the pathname component this also treats the
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
a `+` modifier after it.  In the pathname component this also treats the
```

Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: This is honestly a bit of a judgement call, but here's the reasoning. In a reference work, readers aren't necessarily reading sequentially. I don't like to imply that they need to. The document as a whole still works sequentially if someone choses to read it that way.

File: content/en-US/api/URLPattern.md [hunk 30]
```diff
@@ -0,0 +1,515 @@
...
+const result = pattern.exec({ pathname: "/product/do/some/thing/cool" });
+
+// `do/some/thing/cool`
+result.pathname.groups.action;
+
+// false
+pattern.test({ pathname: "/product" });
+```
+
+### Example: Making matching groups optional and repeated.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Making Matching Groups Optional and Repeated
```

File: content/en-US/api/URLPattern.md [hunk 31]
```diff
@@ -0,0 +1,515 @@
...
+const result = pattern.exec({ pathname: "/product/do/some/thing/cool" });
+
+// `do/some/thing/cool`
+result.pathname.groups.action;
+
+// true
+pattern.test({ pathname: "/product" });
+```
+
+### Example: Using a custom prefix or suffix for an optional or repeated modifier.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Using a Custom Prefix or Suffix for an Optional or Repeated Modifier
```

File: content/en-US/api/URLPattern.md [hunk 32]
```diff
@@ -0,0 +1,515 @@
...
+// true
+pattern.test({ pathname: "/product/" });
+
+const result = pattern.exec({ pathname: "/product/" });
+
+// empty object
+result.pathname.groups
+```
+
+### Example: Using multiple components and features at once.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
### Using Multiple Components and Features at Once
```

**Review by jpmedley — CHANGES_REQUESTED** [review]
Ben,

Thank you for submitting this. One thing I should make clear is that there's nothing inherently wrong with the ways you did much of this. However, writers must always match the style requirements of the site they're publishing on. Most of my requested changes are to bring this in line with what MDN requires. There are a few that request more information. I'm trying to make this as self-contained as it can be for something that's not a complete API reference.

Generally, this is remarkably free of typos, which I really appreciate. I also like the generous inclusion of examples. I don't always get that.

Joe

File: content/en-US/api/URLPattern.md (hunk 1 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md:29]: It can be used for matching URLs in service workers, but no its not only for that.  You can use it for routing within a single page app in a document, for example.  It works in all window and worker javascript contexts.

File: content/en-US/api/URLPattern.md (hunk 2 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md:71]: When a URL is parsed it is automatically normalized to a canonical form.  For example, unicode characters are percent encoded in the pathname, punny encoding in the hostname, default port numbers are elided, paths like `/foo/./bar/` are collapsed to just `/foo/bar`, etc.  URLPattern also implements these changes as well.

In addition, there are some pattern representations that parse to the same underlying meaning.  Like `foo` and `{foo}`.  In those cases we normalize the resulting pattern to the simplest form.  In this case `{foo}` gets changed to `foo`.

File: content/en-US/api/URLPattern.md (hunk 3 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: Yes, the webidl is correct.  I'll update the PR today or tomorrow with the additional information you are requesting here.

File: content/en-US/api/URLPattern.md (hunk 28 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: Did you intend to drop the leading "Example:" here?

File: content/en-US/api/URLPattern.md [hunk 33]
```diff
@@ -0,0 +1,515 @@
...
+
+## Description
+
+// Complete the sentence below with a basic description of the API.
+// You may add more sentences if needed.
+
+The `URLPattern` interface provides a web platform primitive for matching
+URLs.  The pattern syntax is adopted from the popular [path-to-regexp][]
+javascript library and is more ergonomic than using regular expressions.
+
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md:31]:
```suggestion

### Normalization

When a URL is parsed it is automatically normalized to a canonical form. For
example, unicode characters are percent encoded in the `pathname` property, punycode
encoding is used in the hostname, default port numbers are elided, paths like
`/foo/./bar/` are collapsed to just `/foo/bar`, etc. In addition, there are some
pattern representations that parse to the same underlying meaning, like `foo` and
`{foo}`. Such cases are normalized to the simplest form. In this case `{foo}` gets
changed to `foo`.

```

**Review by jpmedley — CHANGES_REQUESTED** [review]
Adding your words to describe normalization.

File: content/en-US/api/URLPattern.md (hunk 3 above)
Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: Leaving open so we remember to check this.

File: content/en-US/api/URLPattern.md (hunk 2 above)
Comment (jpmedley) [reply content/en-US/api/URLPattern.md:71]: I submitted another request for changes to add a description of this to the introduction.

File: content/en-US/api/URLPattern.md (hunk 28 above)
Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: Yes. As with much else, this is how MDN does it.

File: content/en-US/api/URLPattern.md (hunk 11 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: No.  But the pattern itself needs to be in quotes.  Look like I forgot that here.

File: content/en-US/api/URLPattern.md (hunk 12 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: Leaving this one open as a reminder to do all of them.

File: content/en-US/api/URLPattern.md (hunk 14 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: That should be URL "separator".  Like how `https://example.com/` has a colon after the `https` protocol.  Pattern strings also use colon for the beginning of named groups like `:foo`.  Another example of this is `?` which is used before URL search params like `https://example.com/?q=1` and in pattern strings as an optional modifier like `:foo?`.

File: content/en-US/api/URLPattern.md (hunk 11 above)
Comment (jpmedley) [reply content/en-US/api/URLPattern.md]: That's what I meant: the pattern.

File: content/en-US/api/URLPattern.md (hunk 3 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: I pushed a new version with more information here.  I didn't feel what you had was quite right, but tried to follow the style to show that there are two different forms of the constructor.  Happy to change further if necessary.

File: content/en-US/api/URLPattern.md (hunk 11 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: Fixed in latest push

File: content/en-US/api/URLPattern.md (hunk 12 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: Fixed in latest push.

File: content/en-US/api/URLPattern.md (hunk 14 above)
Comment (wanderview) [reply content/en-US/api/URLPattern.md]: I tried to reword this to be more clear in the latest push.

File: content/en-US/api/URLPattern.md [hunk 34]
```diff
@@ -0,0 +1,549 @@
...
+## Constructor
+
+Creates a new `URLPattern` object.  It has two forms.
+
+**`new URLPattern(obj)`**
+
+This form of the constructor takes a `URLPatternInit` dictionary object that
+describes the URLs you want to match.  Its members can be any of `protocol`,
+`username`, `password`, `hostname`, `port`, `pathname`, `search`, `hash`, or
+`baseURL`.  If the `baseURL property is provided it will be parsed as a URL
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
`baseURL`.  If the `baseURL` property is provided it will be parsed as a URL
```

File: content/en-US/api/URLPattern.md [hunk 35]
```diff
@@ -0,0 +1,549 @@
...
+and used to populate any other properties that are missing.  If the `baseURL`
+property is missing, then any other missing properties default to the pattern
+`*` wildcard accepting any input.
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
`*` wildcard, accepting any input.
```

File: content/en-US/api/URLPattern.md [hunk 36]
```diff
@@ -0,0 +1,549 @@
...
+
+**`new URLPattern(pattern, baseURL)`**
+
+This form of the constructor takes a URL string that contains patterns embedded
+in it.  The URL string may be relative if a base URL is provided as the second
+argument.  Note, it may be necessary to escape some characters in the URL
+string where its ambiguous whether the character is separating different URL
+components or if its instead part of a pattern.  For example, you must write
```
Comment (jpmedley) [inline content/en-US/api/URLPattern.md]:
```suggestion
components or if it's instead part of a pattern.  For example, you must write
```

**Review by jpmedley — CHANGES_REQUESTED** [review]
Your new submission looks pretty good. There are just a few typos.

**Comment by wanderview** [comment]
Suggestions applied.  Thanks!
